Experimental change to add type checking to flight map storage.

pshapiro4broad opened the pull request:
It's been bothering me that the flight map is a map without any validation on the content. Additionally, code must be updated whenever the type changes so that the correct type is passed to the `get()` method. Now that `FightMap` supports jackson's `TypeReference`, we can use it to store the type of a map entry to both validate on `put()` and provide the correct type for `get()`.

Start with `JobMapKeys`, `JobMapKeysTest`, then look at some of the `Step` files to see how this change is used. This only covers the keys in `JobMapKeys` but could be extended to all the flight map keys in TDR, probably by adding a common interface for all map keys that contain the `add` and `get` methods.

While doing this I noticed that the keys `REQUEST` and `RESPONSE` use different types for different flights, which means no type checking can be performed. It might be better to have different keys for different flights to enforce types there as well.

I also make a few other unrelated changes, like moving some `String` objects to `UUID`.

pshapiro4broad committed:
Experimental change to add type checking to flight map storage.

pshapiro4broad commented on src/main/java/bio/terra/service/job/JobBuilder.java:
```diff
@@ -23,35 +25,41 @@ public JobBuilder(
 
     return this;
   }
 
```
This can be removed when all keys have been converted from `String` to an object.

pshapiro4broad commented on src/main/java/bio/terra/service/job/JobMapKeys.java:
```diff
@@ -1,29 +1,64 @@
 
-  private String keyName;
+  JobMapKeys() {
+    this(null, new TypeReference<>() {});
```
This generates a `TypeReference` of `Object`

pshapiro4broad commented on src/test/java/bio/terra/service/job/JobMapKeysTest.java:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void testPutTypeCheck() {
```
Need to add a test for `get` type check too; where the correct type is `put` but the wrong type is `get`.

fboulnois commented on src/main/java/bio/terra/service/dataset/flight/datadelete/DatasetDataDeleteFlight.java:
```diff
@@ -35,17 +35,17 @@ public DatasetDataDeleteFlight(FlightMap inputParameters, Object applicationCont
     addStep(
         new VerifyAuthorizationStep(
-            iamClient, IamResourceType.DATASET, datasetId, IamAction.SOFT_DELETE));
+            iamClient, IamResourceType.DATASET, datasetId.toString(), IamAction.SOFT_DELETE));
```
I wonder if there's any way that we could rewrite the `VerifyAuthorizationStep` so that it can also type check the `resourceId` (in this case, the `datasetId`).

pshapiro4broad replied:
I agree, I think we'd benefit from a larger `String` -> `UUID` code review that converted everything to `UUID` where possible. This ends up calling a method on `IamProviderInterface` and none of its APIs accept `UUID`, so that's where I'd start, and propagate the change upwards.

fboulnois commented on src/main/java/bio/terra/service/job/JobBuilder.java:
```diff
@@ -23,35 +25,41 @@ public JobBuilder(
 
+  public JobBuilder addParameter(String key, Object val) {
+    // note that this call overwrites a parameter if it already exists
+    jobParameterMap.put(key, val);
```
What happens if key is `null`? Before there was a null check.

pshapiro4broad replied:
This is placeholder code to get things to compile until all keys are replaced with objects. Even so, I wonder how the `null` error could ever be triggered; the arguments are fixed and aren't based on inputs or any other variables. The underlying code will eventually NPE when it calls `Map.put()` with a null key, so I'm not sure what benefit there is for having an explicit check here.

fboulnois commented on src/main/java/bio/terra/service/profile/flight/update/ProfileUpdateFlight.java:
```diff
@@ -18,16 +18,13 @@ public ProfileUpdateFlight(FlightMap inputParameters, Object applicationContext)
+    AuthenticatedUserRequest user = JobMapKeys.AUTH_USER_INFO.get(inputParameters);
 
-    addStep(new UpdateProfileRetrieveExistingProfileStep(profileService, request, user));
+    addStep(new UpdateProfileRetrieveExistingProfileStep(profileService, request));
```
What happened to the user?

pshapiro4broad replied:
It turned out the step never used it. This was an incidental change I made while looking at the code and isn't part of the map key change.

fboulnois commented on src/test/java/bio/terra/service/dataset/DatasetStorageAccountDaoTest.java:
```diff
@@ -67,7 +63,6 @@ public void setUp() throws Exception {
     BillingProfileRequestModel profileRequest =
         ProfileFixtures.randomizeAzureBillingProfileRequest();
     billingProfile = profileDao.createBillingProfile(profileRequest, "testUser");
-    billingProfileIds.add(billingProfile.getId());
```
What happened to these `billingProfileIds`?

pshapiro4broad replied:
This was never used; the code added to it but never looked at its contents.

fboulnois commented on src/main/java/bio/terra/service/job/JobMapKeys.java:
```diff
@@ -1,29 +1,64 @@
-  AUTH_USER_INFO("auth_user_info"),
-  SUBJECT_ID("subjectId"),
-  CLOUD_PLATFORM("cloudPlatform"),
+  DESCRIPTION("", new TypeReference<>() {}),
+  REQUEST,
+  REVERT_TO(new BillingProfileModel(), new TypeReference<>() {}),
+  RESPONSE,
+  STATUS_CODE(HttpStatus.OK, new TypeReference<>() {}),
+  AUTH_USER_INFO(new AuthenticatedUserRequest(), new TypeReference<>() {}),
+  SUBJECT_ID("", new TypeReference<>() {}),
+  CLOUD_PLATFORM(CloudPlatform.GCP, new TypeReference<>() {}),
```
This part is confusing to me. Perhaps a small comment could be added? In particular, why are the parameter names now blank?

pshapiro4broad replied:
The code now uses the enum names to generate the key names. Since the key names are now private, this seems OK to me; nothing outside the put/get is relying on these names.

Regarding the odd blank or empty parameters, this is how I enabled the type inference for `TypeReference<T>` to work. There has to be something in the constructor that references the type `T` that you want to capture.

While doing this it occurred to me that there's a few cases where, if a key is missing, a default value is used, and it seems like the code could use the argument as the default value in those cases. I didn't look at all uses so I'm not sure it will always work but it seems like a possible usage. Either way I think the code should gather up all the different types into another util class as constants that the code can use here.

fboulnois reviewed (approved):
Overall, the changes seem reasonable to me, I have a few questions though:

pshapiro4broad committed:
- Fix NPE issue with flights that have no request parameter.

- Re-enable testDeleteProfile() by using JobService mock instead of JobBuilder mock.

- Remove unused submitAndWait API

pshapiro4broad committed:
More key-related fixes

- fix issue where a flight was using the wrong key type to get dataset ID
- remove CloudPlatformWrapper.of(String) since it's no longer used
- another String -> UUID type change

pshapiro4broad committed:
Fix issue where a null dataset ID in a test caused a type error. Also made other misc cleanups.

pshapiro4broad committed:
Merge remote-tracking branch 'origin/develop' into ps-working-map-types

# Conflicts:
#	src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestWorkerFlight.java
#	src/test/java/bio/terra/service/filedata/flight/ingest/IngestDriverStepTest.java
